# Team-09 Hack.Eda SF+Y.E

## Условие задачи

Имеется неизвестный нам алгоритм по объединению заказов и результат его работы на выборке ресторанов А. В задаче нужно предсказать поведение алгоритма на выборке ресторанов B. Скачать данные можно на странице с описанием задачи, в формате .xlsx. Это не привычный нам по курсу формат .csv, и смена расширения на .csv здесь не сработает. Поэтому первым делом вам нужно найти способ загрузить эти данные в Python. Рассмотрим по очереди, что означает каждый столбец.

**brand_name**. Это и есть выборка ресторанов (А или В). Все строки с brand_name == А - это реальные обезличенные данные о заказах в Яндекс Еде, при этом работал алгоритм, объединяющий некоторые заказы (не более двух!). Все строки с brand_name == В - это тоже реальные данные о заказах, но в других ресторанах, для которых алгоритм объдинения заказов не работал. Ваша задача - предсказать, какой процент заказов был бы объединен, если бы работал алгоритм объединения. Например, если у нас всего 3 заказа, и первый объединен со вторым, то объединено 66.7% заказов.

**rest_id**. Идентификатор конкретного ресторана (то есть конкретного адреса, откуда курьер забирает заказ). Заказы могут объединяться только если у них один и тот же rest_id.

**order_id**. Уникальный идентификатор заказа.

**batched_with_order_id**. Идентификатор заказа, объединенного с текущим (при наличии). Очевидно, что в выборке В это поле всегда пусто.

**first_in_multiorder_flg**. Является ли заказ первым в мультизаказе (при объединении двух заказов).

**courier_id**. Уникальный идентификатор курьера.

**order_items_cnt**. Количество позиций в заказе.

**city**. Уникальный идентификатор города. Все заказы выборки А принадлежат одному и тому же городу, а вот в выборке В есть разные города.

**order_created_datetime**. Время создания заказа. Обратите внимание, указан 2015 год, но это не значит, что задача старая. Дело в том, что данные обезличены, и все даты тоже изменены.

In [16]:
%% pip install https://github.com/pandas-profiling/pandas-profiling/archive/master.zip

  Created wheel for pandas-profiling: filename=pandas_profiling-2.9.0-py2.py3-none-any.whl size=258957 sha256=8531c44fa13f6d59e619cdd6e2631473ad3825044c9f86d0480ca716282cfd06
  Stored in directory: C:\Users\TANYA\AppData\Local\Temp\pip-ephem-wheel-cache-roqo5br8\wheels\64\b6\85\dfc808b23666a5910371784e349d28818006ff63ed9cfeca59
Successfully built pandas-profiling
  Attempting uninstall: pandas-profiling
    Found existing installation: pandas-profiling 2.4.0
    Uninstalling pandas-profiling-2.4.0:
      Successfully uninstalled pandas-profiling-2.4.0
Note: you may need to restart the kernel to use updated packages.


In [18]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')
warnings.simplefilter('ignore')

print("let's work")

let's work


In [19]:

# загружаем набор данных 
data = pd.read_excel('yandex_hack.xlsx')
data.head()

order_created_datetime brand_name                           rest_id  \
0  2015-09-15 21:08:43.000000          B  7cc98331ddb8ccffe464455f3df763d9   
1  2015-09-13 18:40:12.000000          A  9b73035bb90cd7de4b9c1ff01e44f178   
2  2015-09-14 01:52:40.000000          A  9b73035bb90cd7de4b9c1ff01e44f178   
3  2015-09-13 17:23:52.000000          B  7cc98331ddb8ccffe464455f3df763d9   
4  2015-09-13 11:15:05.000000          B  7cc98331ddb8ccffe464455f3df763d9   

  batched_with_order_id                          order_id  \
0                   NaN  0b6f25c79b35fd1e6d28d845b7ee511d   
1                   NaN  793f2fa9aa3d52d2da05caa3c30cb77e   
2                   NaN  b6c5ec927a3e314b0900c103a0cd68fa   
3                   NaN  5dc658f0928569dbefac74c9f311031c   
4                   NaN  3da09410aeab781baf4b7039fc9b125c   

   first_in_multiorder_flg                        courier_id  order_items_cnt  \
0                      NaN  3f1b96f620988a92d3409ea05e87bd36                3   
1                      NaN  8ed37467dc9b257b3b3bdc1bfebd9b28                8   
2                      NaN  5b17547f825d9cf35c7853621dd842f8                1   
3                      NaN  81647a72700bd46635d9606082360cc8                6   
4                      NaN  065c18216b6d3369d4cc36418c6e80f2                7   

                               city  
0  c6fbacc5cab7e134c9db2a4bbfcc5b64  
1  e6705560085199e9674cd721ee844d0d  
2  e6705560085199e9674cd721ee844d0d  
3  c6fbacc5cab7e134c9db2a4bbfcc5b64  
4  c6fbacc5cab7e134c9db2a4bbfcc5b64

In [20]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66464 entries, 0 to 66463
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   order_created_datetime   66464 non-null  object 
 1   brand_name               66464 non-null  object 
 2   rest_id                  66464 non-null  object 
 3   batched_with_order_id    442 non-null    object 
 4   order_id                 66464 non-null  object 
 5   first_in_multiorder_flg  442 non-null    float64
 6   courier_id               66202 non-null  object 
 7   order_items_cnt          66464 non-null  int64  
 8   city                     66464 non-null  object 
dtypes: float64(1), int64(1), object(7)
memory usage: 4.6+ MB


## Разведовательный анализ данных

In [21]:
import pandas_profiling
pandas_profiling.ProfileReport(data)

KeyError: 'Requested level (var1) does not match index name (None)'

In [22]:
# Преобразуем формат признака
data['order_created_datetime'] = pd.to_datetime(data['order_created_datetime'])

In [23]:
data = data.sort_values('order_created_datetime').reset_index(drop=True)

In [24]:
# Мультизаказы есть только в одном городе
city = data[(data['brand_name'] == 'A')&(data['batched_with_order_id'].notna())]['city'].unique()
city

array(['e6705560085199e9674cd721ee844d0d'], dtype=object)

In [25]:
a = data[(data['brand_name'] == 'A')&(data['city'].isin(city))]
b = pd.DataFrame(data[data['brand_name'] == 'B'])

In [26]:
a.info()
# DataFrame a содержит только заказы ресторанов А в одном городе (где производились тесты)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4878 entries, 0 to 66445
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   order_created_datetime   4878 non-null   datetime64[ns]
 1   brand_name               4878 non-null   object        
 2   rest_id                  4878 non-null   object        
 3   batched_with_order_id    442 non-null    object        
 4   order_id                 4878 non-null   object        
 5   first_in_multiorder_flg  442 non-null    float64       
 6   courier_id               4858 non-null   object        
 7   order_items_cnt          4878 non-null   int64         
 8   city                     4878 non-null   object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(6)
memory usage: 381.1+ KB


## Из 4878 заказов, объединились 442, т.е. процент объединения заказов для группы А = 442 / 4878 = 9.06% 

На основе разведовательного анализа установлено, что вероятность объединения заказа с предыдущим зависит от времени между заказами. 

Если эту плотность вероятности (зависимость вероятности от времени между ближайшими заказами) наложить на выборку ресторанов В, то получим искомый % (прогноз доли мульизаказов в общем объёме заказов).

Выразим вероятность объединения заказов в зависимости от времени между ними для группы А.

In [27]:
a.sort_values(['rest_id', 'order_created_datetime']).reset_index(drop=True)

order_created_datetime brand_name                           rest_id  \
0       2015-09-13 11:24:24          A  03e61cfb9d3325c7c88accd2b3cfa2be   
1       2015-09-13 11:35:47          A  03e61cfb9d3325c7c88accd2b3cfa2be   
2       2015-09-13 11:54:31          A  03e61cfb9d3325c7c88accd2b3cfa2be   
3       2015-09-13 11:58:16          A  03e61cfb9d3325c7c88accd2b3cfa2be   
4       2015-09-13 12:16:48          A  03e61cfb9d3325c7c88accd2b3cfa2be   
...                     ...        ...                               ...   
4873    2015-09-14 17:45:38          A  ff3b46f6abdf1fa1b262783ab3d96b55   
4874    2015-09-14 19:03:48          A  ff3b46f6abdf1fa1b262783ab3d96b55   
4875    2015-09-15 15:19:19          A  ff3b46f6abdf1fa1b262783ab3d96b55   
4876    2015-09-15 19:41:30          A  ff3b46f6abdf1fa1b262783ab3d96b55   
4877    2015-09-16 09:12:04          A  ff3b46f6abdf1fa1b262783ab3d96b55   

                 batched_with_order_id                          order_id  \
0                                  NaN  96abdf37ff71f3ab4cbb1cfcb43a9a3d   
1                                  NaN  399ceb9a041df0a338ecdfcccd4309f4   
2     e0c4053a9330db0c071077c417a186fc  87b092f55dca7eafdcfff64c80b22c89   
3     87b092f55dca7eafdcfff64c80b22c89  e0c4053a9330db0c071077c417a186fc   
4                                  NaN  2e34f208af9e41fd66cbf7b9d91c253a   
...                                ...                               ...   
4873                               NaN  cb93db4787c827b5c1a6285d683c42be   
4874                               NaN  b212f6242f41e77239f105cfb86a9a31   
4875                               NaN  e57d1becd75c6bfd95fe5f996a9c7fa2   
4876                               NaN  91bc4405842c2657f1aed3a9e31cc37e   
4877                               NaN  ebdeb481984fa586f1c27e61bfbc80ce   

      first_in_multiorder_flg                        courier_id  \
0                         NaN  2af733d3c98db5226871a158d36d999d   
1                         NaN  2c836ec50a9e78259de02aff97fa0da9   
2                         1.0  4e3243405d91b074e142849272d56ea8   
3                         0.0  4e3243405d91b074e142849272d56ea8   
4                         NaN  2c836ec50a9e78259de02aff97fa0da9   
...                       ...                               ...   
4873                      NaN  a0e2d3e2bf5d0af07d8beef2c959846a   
4874                      NaN  293cc5c6c0e2af7e3e4f03c18ab1ef78   
4875                      NaN  93906090df9a3279dadd258f82f051ae   
4876                      NaN  1cebb1d1da285d7191e4e6a72ccaf32a   
4877                      NaN  5b17547f825d9cf35c7853621dd842f8   

      order_items_cnt                              city  
0                   1  e6705560085199e9674cd721ee844d0d  
1                   6  e6705560085199e9674cd721ee844d0d  
2                   6  e6705560085199e9674cd721ee844d0d  
3                   4  e6705560085199e9674cd721ee844d0d  
4                   5  e6705560085199e9674cd721ee844d0d  
...               ...                               ...  
4873                4  e6705560085199e9674cd721ee844d0d  
4874                2  e6705560085199e9674cd721ee844d0d  
4875                3  e6705560085199e9674cd721ee844d0d  
4876                5  e6705560085199e9674cd721ee844d0d  
4877                9  e6705560085199e9674cd721ee844d0d  

[4878 rows x 9 columns]

In [49]:
# добавляем уникальный id_мультизаказа: 'batch_id' (для дальнейшего удобства работы с мультизаказми)
def get_batch_id(order):
    """
    формируем batch_id как строку из двух IDs заказов: сначала максимальный. потом минимальный
    можно было бы через lambda, но так нагляднее
    """
    batch_id = order['batched_with_order_id']
    if order['batched_with_order_id'] is not None:
        if str(order['order_id']) > str(order['batched_with_order_id']):
            batch_id = str(order['order_id']) + '|' + str(order['batched_with_order_id'])
        else: batch_id = str(order['batched_with_order_id']) + '|' + str(order['order_id'])
    return batch_id

a['batch_id'] = a.apply(get_batch_id, axis=1)

display(a['batch_id'])

a_batched = a[a['batched_with_order_id'].notna()]

0                    nan|96abdf37ff71f3ab4cbb1cfcb43a9a3d
1                    nan|399ceb9a041df0a338ecdfcccd4309f4
2       e0c4053a9330db0c071077c417a186fc|87b092f55dca7...
3       e0c4053a9330db0c071077c417a186fc|87b092f55dca7...
4                    nan|2e34f208af9e41fd66cbf7b9d91c253a
                              ...                        
4873                 nan|cb93db4787c827b5c1a6285d683c42be
4874                 nan|b212f6242f41e77239f105cfb86a9a31
4875                 nan|e57d1becd75c6bfd95fe5f996a9c7fa2
4876                 nan|91bc4405842c2657f1aed3a9e31cc37e
4877                 nan|ebdeb481984fa586f1c27e61bfbc80ce
Name: batch_id, Length: 4878, dtype: object

In [29]:
a_batched.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 442 entries, 257 to 65500
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   order_created_datetime   442 non-null    datetime64[ns]
 1   brand_name               442 non-null    object        
 2   rest_id                  442 non-null    object        
 3   batched_with_order_id    442 non-null    object        
 4   order_id                 442 non-null    object        
 5   first_in_multiorder_flg  442 non-null    float64       
 6   courier_id               442 non-null    object        
 7   order_items_cnt          442 non-null    int64         
 8   city                     442 non-null    object        
 9   batch_id                 442 non-null    object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(7)
memory usage: 38.0+ KB


In [30]:
# a_batched только мультизаказы

In [31]:
a_batched.head()

order_created_datetime brand_name                           rest_id  \
257     2015-09-13 04:00:11          A  6c1ad5c66540b8c174ff7d9307027299   
1795    2015-09-13 09:11:50          A  8aec6c26cf9db4db10ed06e5e39681ed   
1819    2015-09-13 09:12:49          A  8aec6c26cf9db4db10ed06e5e39681ed   
2869    2015-09-13 09:53:28          A  707f1da36335ef7234721b49991d39c3   
3038    2015-09-13 10:00:37          A  707f1da36335ef7234721b49991d39c3   

                 batched_with_order_id                          order_id  \
257   d0febfc6653fa848cf0a9fadf98261dd  931ad84269827a4c27ce66734460cdf7   
1795  04c7d89ab6b8ff72fb5868413602ddef  5ae2e25193da2000e07dc2051b5ca657   
1819  5ae2e25193da2000e07dc2051b5ca657  04c7d89ab6b8ff72fb5868413602ddef   
2869  83044987a923b13056db64220dd548c5  287564b3ea79fef51a657c6a4237deb5   
3038  287564b3ea79fef51a657c6a4237deb5  83044987a923b13056db64220dd548c5   

      first_in_multiorder_flg                        courier_id  \
257                       0.0  a973a8d1969c72a9eb8147d995e519bf   
1795                      1.0  ebe67e247939ab69f7734ba1557b763d   
1819                      0.0  ebe67e247939ab69f7734ba1557b763d   
2869                      0.0  cd30a21c6bbbcc784de70aa8627be584   
3038                      1.0  cd30a21c6bbbcc784de70aa8627be584   

      order_items_cnt                              city  \
257                 7  e6705560085199e9674cd721ee844d0d   
1795                2  e6705560085199e9674cd721ee844d0d   
1819               14  e6705560085199e9674cd721ee844d0d   
2869                8  e6705560085199e9674cd721ee844d0d   
3038                5  e6705560085199e9674cd721ee844d0d   

                                               batch_id  
257   d0febfc6653fa848cf0a9fadf98261dd|931ad84269827...  
1795  5ae2e25193da2000e07dc2051b5ca657|04c7d89ab6b8f...  
1819  5ae2e25193da2000e07dc2051b5ca657|04c7d89ab6b8f...  
2869  83044987a923b13056db64220dd548c5|287564b3ea79f...  
3038  83044987a923b13056db64220dd548c5|287564b3ea79f...

In [52]:
batched_min = a_batched[['batch_id', 'order_created_datetime']].groupby(["batch_id"], as_index=False).min().reset_index()
batched_min['first_second'] = 'first'
batched_max = a_batched[['batch_id', 'order_created_datetime']].groupby(["batch_id"], as_index=False).max().reset_index()
batched_max['first_second'] = 'second'
batched_all = pd.concat([batched_min, batched_max])
batched_all
display(batched_all)
print(batched_all['first_second'].unique())

index                                           batch_id  \
0        0  182abd4ca6e9d2108bb14b49da64bc26|080acf5392740...   
1        1  245ac1abd392bf31a438f66f41fa736d|1803eefbcb115...   
2        2  2bec9832c989ad72518698f223f1a5cc|027ca436004f2...   
3        3  3100c1f8902818f1c1346cedded9de9f|1dd3c43d929c5...   
4        4  342bd95202383b71d74d949c2ba4618f|1b9948e40c868...   
..     ...                                                ...   
216    216  fd6fce20b5faec0ebb4e274dab6865de|e6d307877e40f...   
217    217  fdde761804315e19f19b8f748c5ff537|27bad12eeb0b6...   
218    218  fe06542117770b36af8859368771a334|8b503566ab43a...   
219    219  fe5749efc1d1e6584f80d7d86c00b003|4914b62257431...   
220    220  ff51a456fe40f0e5c5b2e4222049eaa4|037fc172de6bc...   

    order_created_datetime first_second  
0      2015-09-15 19:25:31        first  
1      2015-09-13 17:49:20        first  
2      2015-09-13 13:29:02        first  
3      2015-09-15 20:25:41        first  
4      2015-09-14 10:14:48        first  
..                     ...          ...  
216    2015-09-14 20:36:17       second  
217    2015-09-14 13:18:23       second  
218    2015-09-13 15:17:32       second  
219    2015-09-15 09:15:00       second  
220    2015-09-15 17:57:05       second  

[442 rows x 4 columns]

['first' 'second']


In [33]:
 a = pd.merge(a, batched_all, how='left', on=['batch_id', 'order_created_datetime'])

In [34]:
a.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4878 entries, 0 to 4877
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   order_created_datetime   4878 non-null   datetime64[ns]
 1   brand_name               4878 non-null   object        
 2   rest_id                  4878 non-null   object        
 3   batched_with_order_id    442 non-null    object        
 4   order_id                 4878 non-null   object        
 5   first_in_multiorder_flg  442 non-null    float64       
 6   courier_id               4858 non-null   object        
 7   order_items_cnt          4878 non-null   int64         
 8   city                     4878 non-null   object        
 9   batch_id                 4878 non-null   object        
 10  index                    442 non-null    float64       
 11  first_second             442 non-null    object        
dtypes: datetime64[ns](1), float64(2), 

In [35]:
# теперь для каждого заказа вычислим время до пердыдущего заказа (с точностью до минут)
# DataSet _a_ отсортирован по rest_id и времени создания заказа
a = a.sort_values(['rest_id', 'order_created_datetime']).reset_index(drop=True)
orders_time = {'order_id':[],'time_shift':[]}
old_rest_id = None
old_order_time = None
for idx, row in a[['rest_id', 'order_id', 'order_created_datetime']].iterrows():
    if old_rest_id == row['rest_id']:
        time_shift = (row['order_created_datetime'] - old_order_time).seconds // 60
    else:
        old_rest_id = row['rest_id']
        time_shift = None
    old_order_time = row['order_created_datetime']
    orders_time['order_id'].append(row['order_id'])
    orders_time['time_shift'].append(time_shift)
print(len(orders_time))        

# конвертируем словарь в DataFrame
order_time_df = pd.DataFrame(orders_time)
# order_time_df['time_shift'].astype(int)
order_time_df.info()
order_time_df.head()

2
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4878 entries, 0 to 4877
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   order_id    4878 non-null   object 
 1   time_shift  4672 non-null   float64
dtypes: float64(1), object(1)
memory usage: 76.3+ KB


order_id  time_shift
0  96abdf37ff71f3ab4cbb1cfcb43a9a3d         NaN
1  399ceb9a041df0a338ecdfcccd4309f4        11.0
2  87b092f55dca7eafdcfff64c80b22c89        18.0
3  e0c4053a9330db0c071077c417a186fc         3.0
4  2e34f208af9e41fd66cbf7b9d91c253a        18.0

In [36]:
# добавляем (join) time_shift к основному набору данных
a = pd.merge(a, order_time_df, how='left', on=['order_id'])
a.head()

order_created_datetime brand_name                           rest_id  \
0    2015-09-13 11:24:24          A  03e61cfb9d3325c7c88accd2b3cfa2be   
1    2015-09-13 11:35:47          A  03e61cfb9d3325c7c88accd2b3cfa2be   
2    2015-09-13 11:54:31          A  03e61cfb9d3325c7c88accd2b3cfa2be   
3    2015-09-13 11:58:16          A  03e61cfb9d3325c7c88accd2b3cfa2be   
4    2015-09-13 12:16:48          A  03e61cfb9d3325c7c88accd2b3cfa2be   

              batched_with_order_id                          order_id  \
0                               NaN  96abdf37ff71f3ab4cbb1cfcb43a9a3d   
1                               NaN  399ceb9a041df0a338ecdfcccd4309f4   
2  e0c4053a9330db0c071077c417a186fc  87b092f55dca7eafdcfff64c80b22c89   
3  87b092f55dca7eafdcfff64c80b22c89  e0c4053a9330db0c071077c417a186fc   
4                               NaN  2e34f208af9e41fd66cbf7b9d91c253a   

   first_in_multiorder_flg                        courier_id  order_items_cnt  \
0                      NaN  2af733d3c98db5226871a158d36d999d                1   
1                      NaN  2c836ec50a9e78259de02aff97fa0da9                6   
2                      1.0  4e3243405d91b074e142849272d56ea8                6   
3                      0.0  4e3243405d91b074e142849272d56ea8                4   
4                      NaN  2c836ec50a9e78259de02aff97fa0da9                5   

                               city  \
0  e6705560085199e9674cd721ee844d0d   
1  e6705560085199e9674cd721ee844d0d   
2  e6705560085199e9674cd721ee844d0d   
3  e6705560085199e9674cd721ee844d0d   
4  e6705560085199e9674cd721ee844d0d   

                                            batch_id  index first_second  \
0               nan|96abdf37ff71f3ab4cbb1cfcb43a9a3d    NaN          NaN   
1               nan|399ceb9a041df0a338ecdfcccd4309f4    NaN          NaN   
2  e0c4053a9330db0c071077c417a186fc|87b092f55dca7...  177.0        first   
3  e0c4053a9330db0c071077c417a186fc|87b092f55dca7...  177.0       second   
4               nan|2e34f208af9e41fd66cbf7b9d91c253a    NaN          NaN   

   time_shift  
0         NaN  
1        11.0  
2        18.0  
3         3.0  
4        18.0

In [53]:
all_orders_by_timeshift = a[['time_shift', 'order_id']].groupby(["time_shift"], as_index=False).count().reset_index()
all_orders_by_timeshift = all_orders_by_timeshift.rename(columns={"order_id": "orders_count"})[['time_shift', 'orders_count']]
all_orders_by_timeshift

time_shift  orders_count
0           0.0            73
1           1.0            67
2           2.0            82
3           3.0            73
4           4.0            80
..          ...           ...
710      1390.0             1
711      1401.0             1
712      1413.0             2
713      1423.0             1
714      1426.0             1

[715 rows x 2 columns]

In [54]:
second_orders_by_timeshift = a[['time_shift', 'order_id']][a['first_second'] == 'second'].groupby(["time_shift"], as_index=False).count().reset_index()
second_orders_by_timeshift = second_orders_by_timeshift.rename(columns={"order_id": "second_orders"})[['time_shift', 'second_orders']]
second_orders_by_timeshift

time_shift  second_orders
0          0.0             20
1          1.0             16
2          2.0             20
3          3.0             17
4          4.0             23
5          5.0             17
6          6.0             16
7          7.0             16
8          8.0             16
9          9.0             14
10        10.0              8
11        11.0              6
12        12.0              6
13        13.0              9
14        14.0              3
15        15.0              4
16        16.0              1
17        17.0              2
18        18.0              1
19        19.0              1
20        20.0              2
21        22.0              1
22        34.0              1
23       519.0              1

In [55]:
a_by_timeshift_df = pd.merge(all_orders_by_timeshift, second_orders_by_timeshift, how = 'outer', on=['time_shift'])
a_by_timeshift_df['second_orders'] = a_by_timeshift_df['second_orders'].fillna(0)
a_by_timeshift_df

time_shift  orders_count  second_orders
0           0.0            73           20.0
1           1.0            67           16.0
2           2.0            82           20.0
3           3.0            73           17.0
4           4.0            80           23.0
..          ...           ...            ...
710      1390.0             1            0.0
711      1401.0             1            0.0
712      1413.0             2            0.0
713      1423.0             1            0.0
714      1426.0             1            0.0

[715 rows x 3 columns]

In [56]:
def get_probability(ts_row):
    probability = 0
    if ts_row['orders_count'] is not None and ts_row['second_orders'] is not None:
        probability = ts_row['second_orders'] / ts_row['orders_count']
    return probability

a_by_timeshift_df['probability'] = a_by_timeshift_df.apply(get_probability, axis=1)
probability_by_timeshift = a_by_timeshift_df[['time_shift', 'probability']]
probability_by_timeshift.head(15)
# при увеличении временного интервала между заказами - вероятность объединения падает. 

time_shift  probability
0          0.0     0.273973
1          1.0     0.238806
2          2.0     0.243902
3          3.0     0.232877
4          4.0     0.287500
5          5.0     0.253731
6          6.0     0.280702
7          7.0     0.225352
8          8.0     0.235294
9          9.0     0.254545
10        10.0     0.150943
11        11.0     0.100000
12        12.0     0.122449
13        13.0     0.157895
14        14.0     0.051724

In [41]:
# Т.е. если второй заказ пришел через столько-то секунд после первого, 
# то вероятность их объединения probability_by_timeshift(time_shift)

In [42]:
# теперь берем DataFrame с ресторанами B, для каждого заказа аналогично вычисляем время в минутах до предыдущего и 
# умножим это время на вероятность, просуммируем полученные значения, умножим на два (т.к. в мультизаказе два заказа) 
# и просуммируем - это будет количество заказов, доставляемых одновременно. 
# Разделим на общее количество заказов и получим искомую величину метрики

In [43]:
# итак, есть DataFrame _B_ содержащий только рестораны В
b.info()
b['brand_name'].unique()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 56835 entries, 3 to 66463
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   order_created_datetime   56835 non-null  datetime64[ns]
 1   brand_name               56835 non-null  object        
 2   rest_id                  56835 non-null  object        
 3   batched_with_order_id    0 non-null      object        
 4   order_id                 56835 non-null  object        
 5   first_in_multiorder_flg  0 non-null      float64       
 6   courier_id               56765 non-null  object        
 7   order_items_cnt          56835 non-null  int64         
 8   city                     56835 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(6)
memory usage: 4.3+ MB


array(['B'], dtype=object)

In [44]:
# сортируем по ресторану и времени заказа.. 
b = b.sort_values(['rest_id', 'order_created_datetime']).reset_index(drop=True)
b.head()

order_created_datetime brand_name                           rest_id  \
0    2015-09-13 08:38:38          B  005dab15a06781cf716d477dfbb6a08f   
1    2015-09-13 09:23:22          B  005dab15a06781cf716d477dfbb6a08f   
2    2015-09-13 09:50:44          B  005dab15a06781cf716d477dfbb6a08f   
3    2015-09-13 09:53:46          B  005dab15a06781cf716d477dfbb6a08f   
4    2015-09-13 10:12:44          B  005dab15a06781cf716d477dfbb6a08f   

  batched_with_order_id                          order_id  \
0                   NaN  3267852ec0189a55587e8417954a6eea   
1                   NaN  85f8d3171bc795e2ac699dbc4ab13aa7   
2                   NaN  db482fdb78173ddd5ecf13f4b3acce3b   
3                   NaN  50ee0573db2bc2270247e7d21a744d7e   
4                   NaN  adba2ee94cd859e4b65ba16657282669   

   first_in_multiorder_flg                        courier_id  order_items_cnt  \
0                      NaN  b5d4b8292b83f38fd7ea22a7c666ab8d                2   
1                      NaN  db283da01a0bf08541afe689d5528b6a                3   
2                      NaN  13b5e6b7cf3bc36ecff9c7192b674379                4   
3                      NaN  db283da01a0bf08541afe689d5528b6a                4   
4                      NaN  db283da01a0bf08541afe689d5528b6a                2   

                               city  
0  e6705560085199e9674cd721ee844d0d  
1  e6705560085199e9674cd721ee844d0d  
2  e6705560085199e9674cd721ee844d0d  
3  e6705560085199e9674cd721ee844d0d  
4  e6705560085199e9674cd721ee844d0d

In [45]:
# вычисляем для каждого заказа время до предыдущего. 
orders_time = {'order_id':[],'time_shift':[]}
old_rest_id = None
old_order_time = None
for idx, row in b[['rest_id', 'order_id', 'order_created_datetime']].iterrows():
    if old_rest_id == row['rest_id']:
        time_shift = (row['order_created_datetime'] - old_order_time).seconds // 60
    else:
        old_rest_id = row['rest_id']
        time_shift = None
    old_order_time = row['order_created_datetime']
    orders_time['order_id'].append(row['order_id'])
    orders_time['time_shift'].append(time_shift)
print(len(orders_time))

# конвертируем словарь в DataFrame
order_time_df = pd.DataFrame(orders_time)
# order_time_df['time_shift'].astype(int)
order_time_df.info()
order_time_df.head()

2
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56835 entries, 0 to 56834
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   order_id    56835 non-null  object 
 1   time_shift  56255 non-null  float64
dtypes: float64(1), object(1)
memory usage: 888.2+ KB


order_id  time_shift
0  3267852ec0189a55587e8417954a6eea         NaN
1  85f8d3171bc795e2ac699dbc4ab13aa7        44.0
2  db482fdb78173ddd5ecf13f4b3acce3b        27.0
3  50ee0573db2bc2270247e7d21a744d7e         3.0
4  adba2ee94cd859e4b65ba16657282669        18.0

In [46]:
# добавляем (join) time_shift к основному набору данных
b = pd.merge(b, order_time_df, how='left', on=['order_id'])
b.head()

order_created_datetime brand_name                           rest_id  \
0    2015-09-13 08:38:38          B  005dab15a06781cf716d477dfbb6a08f   
1    2015-09-13 09:23:22          B  005dab15a06781cf716d477dfbb6a08f   
2    2015-09-13 09:50:44          B  005dab15a06781cf716d477dfbb6a08f   
3    2015-09-13 09:53:46          B  005dab15a06781cf716d477dfbb6a08f   
4    2015-09-13 10:12:44          B  005dab15a06781cf716d477dfbb6a08f   

  batched_with_order_id                          order_id  \
0                   NaN  3267852ec0189a55587e8417954a6eea   
1                   NaN  85f8d3171bc795e2ac699dbc4ab13aa7   
2                   NaN  db482fdb78173ddd5ecf13f4b3acce3b   
3                   NaN  50ee0573db2bc2270247e7d21a744d7e   
4                   NaN  adba2ee94cd859e4b65ba16657282669   

   first_in_multiorder_flg                        courier_id  order_items_cnt  \
0                      NaN  b5d4b8292b83f38fd7ea22a7c666ab8d                2   
1                      NaN  db283da01a0bf08541afe689d5528b6a                3   
2                      NaN  13b5e6b7cf3bc36ecff9c7192b674379                4   
3                      NaN  db283da01a0bf08541afe689d5528b6a                4   
4                      NaN  db283da01a0bf08541afe689d5528b6a                2   

                               city  time_shift  
0  e6705560085199e9674cd721ee844d0d         NaN  
1  e6705560085199e9674cd721ee844d0d        44.0  
2  e6705560085199e9674cd721ee844d0d        27.0  
3  e6705560085199e9674cd721ee844d0d         3.0  
4  e6705560085199e9674cd721ee844d0d        18.0

In [59]:
b_with_probability = pd.merge(b, probability_by_timeshift, how='left', on=['time_shift'])
display(b_with_probability[b_with_probability.order_id == '50ee0573db2bc2270247e7d21a744d7e'])
b_forecast = b_with_probability[['order_created_datetime', 'city', 'rest_id', 'order_id', 'time_shift', 'probability']]
b_forecast['probability'] = b_forecast['probability'].fillna(0)
# умножаем на 2 т.к. один заказт добавляется к предыдущему и два заказа едут вместе, а не один
b_forecast['joined'] = b_forecast['probability'] * 2

b_forecast.head()

order_created_datetime brand_name                           rest_id  \
3    2015-09-13 09:53:46          B  005dab15a06781cf716d477dfbb6a08f   

  batched_with_order_id                          order_id  \
3                   NaN  50ee0573db2bc2270247e7d21a744d7e   

   first_in_multiorder_flg                        courier_id  order_items_cnt  \
3                      NaN  db283da01a0bf08541afe689d5528b6a                4   

                               city  time_shift  probability  
3  e6705560085199e9674cd721ee844d0d         3.0     0.232877

order_created_datetime                              city  \
0    2015-09-13 08:38:38  e6705560085199e9674cd721ee844d0d   
1    2015-09-13 09:23:22  e6705560085199e9674cd721ee844d0d   
2    2015-09-13 09:50:44  e6705560085199e9674cd721ee844d0d   
3    2015-09-13 09:53:46  e6705560085199e9674cd721ee844d0d   
4    2015-09-13 10:12:44  e6705560085199e9674cd721ee844d0d   

                            rest_id                          order_id  \
0  005dab15a06781cf716d477dfbb6a08f  3267852ec0189a55587e8417954a6eea   
1  005dab15a06781cf716d477dfbb6a08f  85f8d3171bc795e2ac699dbc4ab13aa7   
2  005dab15a06781cf716d477dfbb6a08f  db482fdb78173ddd5ecf13f4b3acce3b   
3  005dab15a06781cf716d477dfbb6a08f  50ee0573db2bc2270247e7d21a744d7e   
4  005dab15a06781cf716d477dfbb6a08f  adba2ee94cd859e4b65ba16657282669   

   time_shift  probability    joined  
0         NaN     0.000000  0.000000  
1        44.0     0.000000  0.000000  
2        27.0     0.000000  0.000000  
3         3.0     0.232877  0.465753  
4        18.0     0.019231  0.038462

In [60]:
# количетсво заказов в ресторанах _B_ 
b_orders_count = b_forecast['order_id'].count()
# количество объедененных заказов:
b_joined_count = int(b_forecast['joined'].sum())
print(f"количетсво заказов в ресторанах _B_: {b_orders_count}, прогноз количества объединенных заказов: {b_joined_count}")
print(f"\nИскомый прогноз % совмещения заказов в ресторанах _В_: {100*b_joined_count/b_orders_count:0.2f}%.")

количетсво заказов в ресторанах _B_: 56835, прогноз количества объединенных заказов: 15075

Искомый прогноз % совмещения заказов в ресторанах _В_: 26.52%.


# Результат: искомый прогноз % совмещения заказов в ресторанах _В_: 26.52%.

# Проверим для ресторанов А

In [32]:
a

order_created_datetime brand_name                           rest_id  \
0       2015-09-13 11:24:24          A  03e61cfb9d3325c7c88accd2b3cfa2be   
1       2015-09-13 11:35:47          A  03e61cfb9d3325c7c88accd2b3cfa2be   
2       2015-09-13 11:54:31          A  03e61cfb9d3325c7c88accd2b3cfa2be   
3       2015-09-13 11:58:16          A  03e61cfb9d3325c7c88accd2b3cfa2be   
4       2015-09-13 12:16:48          A  03e61cfb9d3325c7c88accd2b3cfa2be   
...                     ...        ...                               ...   
4873    2015-09-14 17:45:38          A  ff3b46f6abdf1fa1b262783ab3d96b55   
4874    2015-09-14 19:03:48          A  ff3b46f6abdf1fa1b262783ab3d96b55   
4875    2015-09-15 15:19:19          A  ff3b46f6abdf1fa1b262783ab3d96b55   
4876    2015-09-15 19:41:30          A  ff3b46f6abdf1fa1b262783ab3d96b55   
4877    2015-09-16 09:12:04          A  ff3b46f6abdf1fa1b262783ab3d96b55   

                 batched_with_order_id                          order_id  \
0                                  NaN  96abdf37ff71f3ab4cbb1cfcb43a9a3d   
1                                  NaN  399ceb9a041df0a338ecdfcccd4309f4   
2     e0c4053a9330db0c071077c417a186fc  87b092f55dca7eafdcfff64c80b22c89   
3     87b092f55dca7eafdcfff64c80b22c89  e0c4053a9330db0c071077c417a186fc   
4                                  NaN  2e34f208af9e41fd66cbf7b9d91c253a   
...                                ...                               ...   
4873                               NaN  cb93db4787c827b5c1a6285d683c42be   
4874                               NaN  b212f6242f41e77239f105cfb86a9a31   
4875                               NaN  e57d1becd75c6bfd95fe5f996a9c7fa2   
4876                               NaN  91bc4405842c2657f1aed3a9e31cc37e   
4877                               NaN  ebdeb481984fa586f1c27e61bfbc80ce   

      first_in_multiorder_flg                        courier_id  \
0                         NaN  2af733d3c98db5226871a158d36d999d   
1                         NaN  2c836ec50a9e78259de02aff97fa0da9   
2                         1.0  4e3243405d91b074e142849272d56ea8   
3                         0.0  4e3243405d91b074e142849272d56ea8   
4                         NaN  2c836ec50a9e78259de02aff97fa0da9   
...                       ...                               ...   
4873                      NaN  a0e2d3e2bf5d0af07d8beef2c959846a   
4874                      NaN  293cc5c6c0e2af7e3e4f03c18ab1ef78   
4875                      NaN  93906090df9a3279dadd258f82f051ae   
4876                      NaN  1cebb1d1da285d7191e4e6a72ccaf32a   
4877                      NaN  5b17547f825d9cf35c7853621dd842f8   

      order_items_cnt                              city  \
0                   1  e6705560085199e9674cd721ee844d0d   
1                   6  e6705560085199e9674cd721ee844d0d   
2                   6  e6705560085199e9674cd721ee844d0d   
3                   4  e6705560085199e9674cd721ee844d0d   
4                   5  e6705560085199e9674cd721ee844d0d   
...               ...                               ...   
4873                4  e6705560085199e9674cd721ee844d0d   
4874                2  e6705560085199e9674cd721ee844d0d   
4875                3  e6705560085199e9674cd721ee844d0d   
4876                5  e6705560085199e9674cd721ee844d0d   
4877                9  e6705560085199e9674cd721ee844d0d   

                                               batch_id  index first_second  \
0                  nan|96abdf37ff71f3ab4cbb1cfcb43a9a3d    NaN          NaN   
1                  nan|399ceb9a041df0a338ecdfcccd4309f4    NaN          NaN   
2     e0c4053a9330db0c071077c417a186fc|87b092f55dca7...  177.0        first   
3     e0c4053a9330db0c071077c417a186fc|87b092f55dca7...  177.0       second   
4                  nan|2e34f208af9e41fd66cbf7b9d91c253a    NaN          NaN   
...                                                 ...    ...          ...   
4873               nan|cb93db4787c827b5c1a6285d683c42be    NaN          NaN   
4874               nan|b212f6242f

In [33]:
a_with_probability = pd.merge(a, probability_by_timeshift, how='left', on=['time_shift'])
a_forecast = a_with_probability[['order_created_datetime', 'city', 'rest_id', 'order_id', 'time_shift', 'probability']]
a_forecast['probability'] = a_forecast['probability'].fillna(0)
# умножаем на 2 т.к. один заказт добавляется к предыдущему и два заказа едут вместе, а не один
a_forecast['joined'] = a_forecast['probability'] * 2
a_forecast.head()

order_created_datetime                              city  \
0    2015-09-13 11:24:24  e6705560085199e9674cd721ee844d0d   
1    2015-09-13 11:35:47  e6705560085199e9674cd721ee844d0d   
2    2015-09-13 11:54:31  e6705560085199e9674cd721ee844d0d   
3    2015-09-13 11:58:16  e6705560085199e9674cd721ee844d0d   
4    2015-09-13 12:16:48  e6705560085199e9674cd721ee844d0d   

                            rest_id                          order_id  \
0  03e61cfb9d3325c7c88accd2b3cfa2be  96abdf37ff71f3ab4cbb1cfcb43a9a3d   
1  03e61cfb9d3325c7c88accd2b3cfa2be  399ceb9a041df0a338ecdfcccd4309f4   
2  03e61cfb9d3325c7c88accd2b3cfa2be  87b092f55dca7eafdcfff64c80b22c89   
3  03e61cfb9d3325c7c88accd2b3cfa2be  e0c4053a9330db0c071077c417a186fc   
4  03e61cfb9d3325c7c88accd2b3cfa2be  2e34f208af9e41fd66cbf7b9d91c253a   

   time_shift  probability    joined  
0         NaN     0.000000  0.000000  
1        11.0     0.100000  0.200000  
2        18.0     0.019231  0.038462  
3         3.0     0.232877  0.465753  
4        18.0     0.019231  0.038462

In [34]:
# количетсво заказов в ресторанах _A_ 
a_orders_count = a_forecast['order_id'].count()
# количество объедененных заказов:
a_joined_count = int(a_forecast['joined'].sum())
print(f"количетсво заказов в ресторанах _А_: {a_orders_count}, прогноз количества объединенных заказов: {a_joined_count}")
print(f"\nИскомый прогноз % совмещения заказов в ресторанах _А_: {100*a_joined_count/a_orders_count:0.2f}%.")

количетсво заказов в ресторанах _А_: 4878, прогноз количества объединенных заказов: 442

Искомый прогноз % совмещения заказов в ресторанах _А_: 9.06%.


# Предложение для ЯндексЕды

Если объединять три заказа в один это даст дополнительно до 13 % в объединении заказов на выборке ресторанов В.

В алгоритм также необходимо заложить в явном виде ограничение на сумму количества продуктов в мультизаказе и если в первом заказе много продуктов, то во втором заказе (к нему) должно быть мало продуктов. 

Тогда, если в первом заказе много продуктов, то ресторан его долго делает и за это время поступает второй заказ и если во втором заказе мало продуктов ресторан успеет сделать оба заказа быстро, а курьер успеет донести за время меньше обещенного.